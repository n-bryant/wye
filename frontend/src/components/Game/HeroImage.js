import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";

import styles from "./HeroImage.styles";

export const PLACEHOLDER_PATH = "/steamHeroPlaceholder.jpg";

/**
 * renders a hero image with a placeholder src value
 */
export const HeroImage = props => {
  const classnames = HeroImage.classnames(props);
  const { altText, imageSrc } = props;
  const [hasBrokenSrc, setHasBrokenSrc] = React.useState(false);
  const [loaded, setLoaded] = React.useState(false);

  const onLoad = () => {
    setLoaded(true);
  };
  const onError = () => {
    setHasBrokenSrc(true);
  };

  return (
    <React.Fragment>
      <img
        className={classnames.element("image", {
          hidden: !loaded
        })}
        onError={onError}
        onLoad={onLoad}
        src={hasBrokenSrc ? PLACEHOLDER_PATH : imageSrc}
        alt={altText}
      />
      <img
        className={classnames.element("placeholder", {
          hidden: loaded
        })}
        src={PLACEHOLDER_PATH}
        alt={altText}
      />
    </React.Fragment>
  );
};
HeroImage.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}HeroImage`
);
HeroImage.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    image: PropTypes.string,
    imageHidden: PropTypes.string,
    placeholder: PropTypes.string,
    placeholderHidden: PropTypes.string
  }),
  // alt text for the image
  altText: PropTypes.string,
  // the location of the image to render
  imageSrc: PropTypes.string
};
HeroImage.defaultProps = {
  classes: {},
  altText: "",
  imageSrc: ""
};

export default withStyles(styles)(HeroImage);
