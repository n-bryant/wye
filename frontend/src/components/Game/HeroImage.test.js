import { HeroImage } from "./HeroImage";

describe("HeroImage", () => {
  const props = {
    classes: {
      image: "image",
      imageSquaredBottom: "imageSquaredBottom",
      imageHidden: "imageHidden",
      placeholder: "placeholder",
      placeholderHidden: "placeholderHidden"
    },
    altText: "alt",
    src: "src"
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render successfully", () => {
    const wrapper = shallow(<HeroImage {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should update the src of the image if the image fails to load", () => {
    const hasBrokenSrc = false;
    const setHasBrokenSrc = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [hasBrokenSrc, setHasBrokenSrc]);
    const wrapperWithState = shallow(<HeroImage {...props} />);
    const image = wrapperWithState.findWhere(n =>
      n.hasClass(props.classes.image)
    );
    image.prop("onError")();
    expect(setHasBrokenSrc).toHaveBeenCalledWith(true);
  });

  it("should update the loaded status of the image once the image has loaded", () => {
    const loaded = false;
    const setLoaded = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(() => [loaded, setLoaded]);

    const wrapperWithState = shallow(<HeroImage {...props} />);
    const image = wrapperWithState.findWhere(n =>
      n.hasClass(props.classes.image)
    );
    image.prop("onLoad")();
    expect(setLoaded).toHaveBeenCalledWith(true);
  });
});
