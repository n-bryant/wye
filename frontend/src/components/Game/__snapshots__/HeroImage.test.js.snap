// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroImage should render successfully 1`] = `
<Fragment>
  <img
    alt="alt"
    className="wye-HeroImage__image wye-HeroImage__image--hidden image imageHidden"
    onError={[Function]}
    onLoad={[Function]}
    src=""
  />
  <img
    alt="alt"
    className="wye-HeroImage__placeholder placeholder"
    src="/steamHeroPlaceholder.jpg"
  />
</Fragment>
`;
